saexec/saexec.go

```diff
@@ -0,0 +1,139 @@
+	logEvents   event.FeedOf[[]*types.Log]
+
+	chainConfig *params.ChainConfig
+	db          ethdb.Database
+	stateCache  state.Database
+	// executeScratchSpace MUST NOT be accessed by any methods other than
+	// [Executor.init], [Executor.execute], and [Executor.Close].
+	executeScratchSpace executionScratchSpace
+}
+
+// New constructs and starts a new [Executor]. Call [Executor.Close] to release
+// resources created by this constructor.
+//
+// The last-executed block MAY be the genesis block for an always-SAE chain, the
+// last pre-SAE synchronous block during transition, or the last asynchronously
+// executed block after shutdown and recovery.
+func New(
+	lastExecuted *blocks.Block,
+	chainConfig *params.ChainConfig,
+	db ethdb.Database,
+	triedbConfig *triedb.Config,
+	hooks hook.Points,
+	log logging.Logger,
+) (*Executor, error) {
+	e := &Executor{
+		quit:        make(chan struct{}), // closed by [Executor.Close]
+		done:        make(chan struct{}), // closed by [Executor.processQueue] after `quit` is closed
+		log:         log,
+		hooks:       hooks,
+		queue:       make(chan *blocks.Block, 4096), // arbitrarily sized
+		chainConfig: chainConfig,
+		db:          db,
+		stateCache:  state.NewDatabaseWithConfig(db, triedbConfig),
+	}
+	e.lastExecuted.Store(lastExecuted)
+	if err := e.init(); err != nil {
+		return nil, err
+	}
+
+	go e.processQueue()
+	return e, nil
+}
+
+func (e *Executor) init() error {
+	last := e.lastExecuted.Load()
+	e.gasClock = last.ExecutedByGasTime().Clone()
+
+	root := last.PostExecutionStateRoot()
+	snapConf := snapshot.Config{
+		CacheSize:  128, // MB
+		AsyncBuild: true,
+	}
+	snaps, err := snapshot.New(snapConf, e.db, e.stateCache.TrieDB(), root)
+	if err != nil {
+		return err
+	}
+	statedb, err := state.New(root, e.stateCache, snaps)
+	if err != nil {
+		return err
+	}
+
+	e.executeScratchSpace = executionScratchSpace{
+		snaps:   snaps,
+		statedb: statedb,
+	}
+	return nil
+}
+
+// Close shuts down the [Executor], waits for the currently executing block
+// to complete, and then releases all resources.
+func (e *Executor) Close() {
+	close(e.quit)
+	<-e.done
+
+	snaps := e.executeScratchSpace.snaps
+	snaps.Disable()
+	snaps.Release()
+}
+
+// ChainConfig returns the config originally passed to [New].
+func (e *Executor) ChainConfig() *params.ChainConfig {
+	return e.chainConfig
+}
+
+// StateCache returns caching database underpinning execution.
+func (e *Executor) StateCache() state.Database {
+	return e.stateCache
+}
+
+// LastExecuted returns the last-executed block in a threadsafe manner.
+func (e *Executor) LastExecuted() *blocks.Block {
+	return e.lastExecuted.Load()
+}
+
+// TimeNotThreadsafe returns a clone of the gas clock that times execution. It
+// is only safe to call when all blocks passed to [Executor.Enqueue]
+// have been executed, and is only intended for use in tests.
+func (e *Executor) TimeNotThreadsafe() *gastime.Time {
+	return e.gasClock.Clone()
+}
```

Comment:
Is there a reason this is:
1. Exported
2. Defined in a production file

At least currently, it is only used is the test in this package.